### ISSUE TITLE
Type guard failed when using union typed variable for props

### ISSUE THREAD

**Issue by tianzhich** [microsoft/TypeScript#40429]
**TypeScript Version:**  4.0.2

I can't open TypeScript Playground with 'Nightly' version so I created with the latest.

**Search Terms:**
type guard failed, union props

**Code**

```ts
interface PluginHook {
    onStoreCreated?: (p1: number) => void;
    onModel?: (p2: string) => void;
}

interface MyPlugin extends PluginHook {
    config?: Record<string, unknown>;
}

type Method = keyof PluginHook;

function fn(p: NonNullable<PluginHook[Method]>) { }

const plugins: MyPlugin[] = [];

for (const plugin of plugins) {
    const method: Method = 'onStoreCreated';

    // worked
    if (plugin['onStoreCreated']) {
        fn(plugin['onStoreCreated'])
    }

    // failed
    // error: Argument of type '((p1: number) => void) | undefined' is not assignable to parameter of type '((p1: number) => void) | ((p2: string) => void)'.
    // Type 'undefined' is not assignable to type '((p1: number) => void) | ((p2: string) => void)'.(2345)
    if (plugin[method]) {
        fn(plugin[method])
    }
}
```

**Expected behavior:**
Type guard checking successfully so `plugin[method]` won't be `undefined` or `null`

**Actual behavior:**
Type guard failed to check undefined for `plugin[method]`.

**Playground Link:** 
[Click here](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vCjBg4SCIMAALE2VkYghgghhE8JSiYj0KGHqEMGA6ZBgQEQEAOTodzHR0OBl0CFKw5NTiAG1DfqGuAF0VRTJkTqyQe2RhJNBbAIQpdQDcniMwQsYMwhF8fn8ZrQpgjkrY3hlkHCwMgpA8TEC8VwRgByOiMZhsDjcYkLGjIAD09OQzWYYy4GOAUxE-xAN1JDCYLHYEE4EC4xKe6LpdM23JmfLJgspIupkoxnTpjI2cGAZ3ZmqZ0CgzAEAEEoNhpBBwEjkN1eshiYIROJkJJZAplGpNNpFAAfOpxCA5EBi4nIYAJJrYuC2Wy5EAnM52gi-OBQOC4-y2+0oJ0uiTSORQFze9z+oQiGzICpOUtuX3EgB0GK1ABUenn6qYQ2GI1GCDG4wmkygmHbO47nWJCx6S16G1wK9Pq7WqvWfUvm4IrABmAAsAFZ5ByuSjQVnHpL0tKaLLz7zLyY1XTvJ4gA)

**Related Issues:**

**Comment by MartinJohns** [comment]
Duplicate of #10530 (and many many more). Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

**Comment by tianzhich** [comment]
Thanks, closed this.

**Comment by tianzhich** [comment]
> Duplicate of #10530 (and many many more). Type narrowing does not occur for indexed access forms `e[k]` where `k` is not a literal.

But do you know how to make it work if I coding this way?

**Comment by j-oliveras** [comment]
Save the `plugin[method]` value into a const variable and use it:
```ts
    const v = plugin[method];
    if (v) {
        fn(v)
    }
```

Full code on [playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vCjBg4SCIMAALE2VkYghgghhE8JSiYj0KGHqEMGA6ZBgQEQEAOTodzHR0OBl0CFKw5NTiAG1DfqGuAF0VRTJkTqyQe2RhJNBbAIQpdQDcniMwQsYMwhF8fn8ZrQpgjkrY3hlkHCwMgpA8TEC8VwRgByOiMZhsDjcYkLGjIAD09OQzWYYy4GOAUxE-xAN1JDCYLHYEE4EC4xKe6LpdM23JmfLJgspIupkoxnTpWI0IxRoNxgxMTz0dM5QnUUul1Fl5vVXiAA).

**Comment by tianzhich** [comment]
> Save the `plugin[method]` value into a const variable and use it:
> 
> ```ts
>     const v = plugin[method];
>     if (v) {
>         fn(v)
>     }
> ```
> 
> Full code on [playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vCjBg4SCIMAALE2VkYghgghhE8JSiYj0KGHqEMGA6ZBgQEQEAOTodzHR0OBl0CFKw5NTiAG1DfqGuAF0VRTJkTqyQe2RhJNBbAIQpdQDcniMwQsYMwhF8fn8ZrQpgjkrY3hlkHCwMgpA8TEC8VwRgByOiMZhsDjcYkLGjIAD09OQzWYYy4GOAUxE-xAN1JDCYLHYEE4EC4xKe6LpdM23JmfLJgspIupkoxnTpWI0IxRoNxgxMTz0dM5QnUUul1Fl5vVXiAA).

Many thanks, mate

**Comment by tianzhich** [comment]
I have a more complicated scenario, @j-oliveras  could you help me resolve [this](https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vX3BoeCRkACE4bHSaZBhmAFFEAAsw5IFS1NJo2PjkYghgghhE8JSiYhVBDNHkKQgwaZMBJeqT0ZgQAUEs7vABADk6D8x0dDgZOgIKU5vgDgBtJYAXRULjUmm0JzhbkRnTAwWEQQuVy4ynWm22u2SSz0FBg9QQYGAdDGIBEn2+v3+gOBoP2JHBhmxJhhijIyE6r3syGESVAtgEITZ4KheJlpKFYGQMiGAkGwyUIxo4ygUwQszFdPOlxMtUe8gECNxlFO2uYQkVIsNtB2or2tj591tyAA9D7kI6pFITVwBBstjs2Us8QByOiMZhsDjcGN6b2nR1XNKat3JcHGnFQr224A7QRZ4ie9Ppx7lg7yYunbzp5s0Tx3bxAA). Thanks in advance.

**Comment by tianzhich** [comment]
Reopened for help.

**Comment by j-oliveras** [comment]
In that case I will use the a [Definite Assignment Assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html#definite-assignment-assertions):
```ts
            if (hook) {
                fn(hook!) // see the ! here
            }
```
[Playground](https://www.typescriptlang.org/play?#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vX3BoeCRkACE4bHSaZBhmAFFEAAsw5IFS1NJo2PjkYghgghhE8JSiYhVBDNHkKQgwaZMBJeqT0ZgQAUEs7vABADk6D8x0dDgZOgIKU5vgDgBtJYAXRULjUmm0JzhbkRnTAwWEQQuVy4ynWm22u2SSz0FBg9QQYGAdDGIBEn2+v3+gOBoP2JHBhmxJhhijIyE6r3syGESVAtgEITZ4KheJlpKFYGQMiGAkGwyUIxo4ygUwQszFdPOlxMtUe8gECNxlFO2uYQkVIsNtB2or2tj591tyAA9D7kI6pFITVwBBstjs2Us8QByOiMZhsDjcGN6b2nR1XNKat3JcHGnFQr224A7QRZ4ie9Ppx7lg4AQnkxdO3nTrZonju3iAA)
I don't know if exists a better solution/workaround.

PD: If you have question, better ask on Stackoverflow.

**Comment by RyanCavanaugh** [comment]
This doesn't appear to be describing a bug

**Comment by tianzhich** [comment]
> This doesn't appear to be describing a bug

well, I will also ask on Stackoverflow, maybe it was a bug, see [this](https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBITAFsARtACUyALwA+ZADcCwLgG4qNOgFkCXCOn5DhAJgEBnMFFDZFqjVt0UAvhQqhIsRBRDAE8MHFBkCAAPSBAuWzQsXBBCEnJ9agQ6GGBsCwAlCCyoLgAee0cQbAAaZEwQYhACAHcQFT1vX3BoeCRkACE4bHSaZBhmAFFEAAsw5IFS1NJo2PjkYghgghhE8JSiYhVBDNHkKQgwaZMBJeqT0ZgQAUEs7vABADk6D8x0dDgZOgIKU5vgDgBtJYAXRULjUmm0JzhbkRnTAwWEQQuVy4ynWm22u2SSz0FBg9QQYGAdDGIBEn2+v3+gOBoP2JHBhmxJhhijIyE6r3syGESVAtgEITZ4KheJlpKFYGQMiGAkGwyUIxo4ygUwQszFdPOlxMtUe8gECNxlFO2uYQkVIsNtB2or2tj591tyAA9D7kI6pFITVwBBstjs2Us8QByOiMZhsDjcGN6b2nR1XNKat3JcHGnFQr224A7QRZ4ie9Ppx7lg7yYunbzp5s0Tx3bxAA) please.

**Comment by j-oliveras** [comment]
@RyanCavanaugh In that case probably is a bug or design limitation (I'm not sure):
```ts
interface PluginHook {
    onStoreCreated?: (p1: string) => void;
    onModel?: (p2: string) => void;
}

declare const plugins: PluginHook[];

function func<Hook extends keyof PluginHook>(method: Hook, fn: (p: NonNullable<PluginHook[keyof PluginHook]>) => void): void {
    for (const plugin of plugins) {
        const hook = plugin[method];
        if (typeof hook !== "undefined") {
            fn(hook);
            // ^^^^
            /*
            Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
              Type '((p1: string) => void) | ((p2: string) => void) | undefined' is not assignable to type '((p1: string) => void) | ((p2: string) => void)'.
                Type 'undefined' is not assignable to type '((p1: string) => void) | ((p2: string) => void)'.
                  Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
                    Type '((p1: string) => void) | ((p2: string) => void) | undefined' is not assignable to type '(p2: string) => void'.
                      Type 'undefined' is not assignable to type '(p2: string) => void'.
            */
        }
        if (hook) {
            fn(hook);
            // ^^^^
            /*
            Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
              Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
            */
        }
        if (!!hook) {
            fn(hook);
            // ^^^^
            /*
            Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
              Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
            */
        }
    }
}
```
**Expected**:
It works

**Actual**:
It say is not compatible. Changing the order of the `if` display the same errors in the exact order. The first error says that can be undefined just after checking for undefined.

[Playground](https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBAZzBRQ2AJTIAvAD5kANwLAuAbio06AWQJcI6fkOEAmSdNkKV6zTooBfChRMJ0cFsgR0pZGEsXBAJAQwcfCJiAG0AXV0KGEwQBDBgOmQUtIAeQhJkCAAPSBAuCWRiCABPAhg0EOiSZUEAWwgwAAtjAQLiABpskAERAQA5OnHMdG8AI3QIXMjQ-tjquoaV5uJ45XtVDS05ASOucj1qGGYhfzCwIKaQWgbgqLCFShpvvwCHnsKike71iHW6xkSlx+wAagjANWEEHqyABpAAhIogQAiVImGCgCBcLGfKE-GgwECCVFyXRkskAenpyAAeqzmaS6fSAFQcskAQSg2EwHXAL2Q8MRyAA5NsQGt+vEpchgJUQAQHnAJBJgNgQHAFigmEEfHAwdAxRKUFLBCJxMgpDIQPIlIcnAoAD5CETWe22J0HRzHKUAOl5dIAKgirTaxDZHc6HGcPV6rHG7C7A1xk7iIPiQISlSrkGqNVqdXqDeKCOKo9KY3aHenE27kJ6Yz7G-6M0mQ2G6chI5KpTm8wXlar1chNdrdfrFlWa0P62mu83jq2Ux2-QnXUHQ-2DwPazKnvKYorx8XJ9Py3PDdXLXXU774wGzr3D4fB9HbSud5nk3bP83xbT0RwJLhCwnUsZwrecjUfa1n07f9333T8v2PcD80gy8SynMtZ0rBDj29YDuycD9Py5ekwzcfsYSEakLgwikqRiGk+xoRkWTZLjqG5fjkAFIURQeZFENlM8SAvIt8JvIj4OrYQTTNKALVI38XybXcsw3IDtNXXS5CojDv2lKSYliBUoKvGDb2Ih9SOQ7cQK0UyDxouiw0YwQ0TRZivk-NjqVpT8eLZdkMMEjCROFCBRQk49LJIazz1s+TCLg+9jSgU1OnNJKly0lC3L0tsyMM1C3Q8zChxSuIbLw68srvBdEMq0qKPc9DDy8-t6JoNw3CAA)

**Comment by tianzhich** [comment]
It was strange, I asked on StackOverflow and received a comment saying it works by using nullish coalescing operator like this: `const hook = plugin[method] ?? undefined`. But we don't know why. Full code on [Playground](https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBAZzBRQ2AJTIAvAD5kANwLAuAbio06AWQJcI6fkOEAmSdNkKV6zTooBfChRMJ0cFsgR0pZGEsXBAJAQwcfCJiAG0AXV0KGEwQBDBgOmQUtIAeQhJkCAAPSBAuCWRiCABPAhg0EOiSZUEAWwgwAAtjAQLiABpskAERAQA5OnHMdG8AI3QIXMjQ-tjquoaV5uJ45XtVDS05ASOucj1qGGYhfzCwIKaQWgbgqLCFShpvvwCHnsKike71iHW6xniyB4PGQqRMMFAEGcPxowAagjANWEEHqyABpAAhIogQAiOEQBEgJEkz6XFE0GAgQT4uS6ek-AD0HOQAD0+Ty6eyOQAqQX0gCCUGwmA64BeyEx2OQAHJtiA1v14srkMBKiACA84BIJMBsCA4AsUEwgj44GDoPLFShlYIROJkFIZCB5EpDk4FAAfIQiawe2zeg6OY7KgB0YvZABUsc7XWIbF6fQ4zoHg1Z03ZfVGuDnyZSkdrdch9YbjabzZaFQQFcmVan3Z6C1n-cgg6nQx2I4Xs7H4+zkEmlcrS4iuBW9QbkEaTWaLYtG83J2384Ou8ce7n++HM37o3Gx2OJ861RqYlqdfOa8v62vrU7W3mwxnI2cR+fz5fWzdbdjyLHM+2A79uyDacqVne8qwXJc61XK0mzfF0PwHECfzPP9-xbKdygpGc5wQx9kIbV8CJDCChycX8-2FDlRzcdk0SEFkLjwxlmRiVlR2+LleX5ASaBFUTqElaVZQeXF0OvGJYk1Ujq0XWsV0opthFte0oEdaigM-TsT2LfdwKMncTLkBi8IA1UnhvEg70rVSkI0l80OozCj0grQbLHJiWNHdjBAJAlOK+P8eJZNk-yE-kBTw8S8KkmUIDlOSCIUkglNvFTEPU59UJtKA7U6B1Ms3QysN80zexoizsP9fz8MnbK4mU+DXMKlD13QhqarovzcPPQL2VYmg3DcIA)

**Comment by tianzhich** [comment]
That was like non-null assertion operator `!` or nullish coalescing operator `??` construct a NonNullable type, so it was compatible with the param which was `NonNullable<T>`. I will use `!`.

**Comment by tianzhich** [comment]
> @RyanCavanaugh In that case probably is a bug or design limitation (I'm not sure):
> 
> ```ts
> interface PluginHook {
>     onStoreCreated?: (p1: string) => void;
>     onModel?: (p2: string) => void;
> }
> 
> declare const plugins: PluginHook[];
> 
> function func<Hook extends keyof PluginHook>(method: Hook, fn: (p: NonNullable<PluginHook[keyof PluginHook]>) => void): void {
>     for (const plugin of plugins) {
>         const hook = plugin[method];
>         if (typeof hook !== "undefined") {
>             fn(hook);
>             // ^^^^
>             /*
>             Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
>               Type '((p1: string) => void) | ((p2: string) => void) | undefined' is not assignable to type '((p1: string) => void) | ((p2: string) => void)'.
>                 Type 'undefined' is not assignable to type '((p1: string) => void) | ((p2: string) => void)'.
>                   Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
>                     Type '((p1: string) => void) | ((p2: string) => void) | undefined' is not assignable to type '(p2: string) => void'.
>                       Type 'undefined' is not assignable to type '(p2: string) => void'.
>             */
>         }
>         if (hook) {
>             fn(hook);
>             // ^^^^
>             /*
>             Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
>               Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
>             */
>         }
>         if (!!hook) {
>             fn(hook);
>             // ^^^^
>             /*
>             Argument of type 'PluginHook[Hook]' is not assignable to parameter of type '((p1: string) => void) | ((p2: string) => void)'.
>               Type 'PluginHook[Hook]' is not assignable to type '(p2: string) => void'.
>             */
>         }
>     }
> }
> ```
> 
> **Expected**:
> It works
> 
> **Actual**:
> It say is not compatible. Changing the order of the `if` display the same errors in the exact order. The first error says that can be undefined just after checking for undefined.
> 
> [Playground](https://www.typescriptlang.org/play?ts=4.0.2#code/JYOwLgpgTgZghgYwgAgAoBsCuBzUAJAewIGtkBvAKGWuQJAGUwCoIBhFuSAEwH4AuZAAoADgEYBAZzBRQ2AJTIAvAD5kANwLAuAbio06AWQJcI6fkOEAmSdNkKV6zTooBfChRMJ0cFsgR0pZGEsXBAJAQwcfCJiAG0AXV0KGEwQBDBgOmQUtIAeQhJkCAAPSBAuCWRiCABPAhg0EOiSZUEAWwgwAAtjAQLiABpskAERAQA5OnHMdG8AI3QIXMjQ-tjquoaV5uJ45XtVDS05ASOucj1qGGYhfzCwIKaQWgbgqLCFShpvvwCHnsKike71iHW6xkSlx+wAagjANWEEHqyABpAAhIogQAiVImGCgCBcLGfKE-GgwECCVFyXRkskAenpyAAeqzmaS6fSAFQcskAQSg2EwHXAL2Q8MRyAA5NsQGt+vEpchgJUQAQHnAJBJgNgQHAFigmEEfHAwdAxRKUFLBCJxMgpDIQPIlIcnAoAD5CETWe22J0HRzHKUAOl5dIAKgirTaxDZHc6HGcPV6rHG7C7A1xk7iIPiQISlSrkGqNVqdXqDeKCOKo9KY3aHenE27kJ6Yz7G-6M0mQ2G6chI5KpTm8wXlar1chNdrdfrFlWa0P62mu83jq2Ux2-QnXUHQ-2DwPazKnvKYorx8XJ9Py3PDdXLXXU774wGzr3D4fB9HbSud5nk3bP83xbT0RwJLhCwnUsZwrecjUfa1n07f9333T8v2PcD80gy8SynMtZ0rBDj29YDuycD9Py5ekwzcfsYSEakLgwikqRiGk+xoRkWTZLjqG5fjkAFIURQeZFENlM8SAvIt8JvIj4OrYQTTNKALVI38XybXcsw3IDtNXXS5CojDv2lKSYliBUoKvGDb2Ih9SOQ7cQK0UyDxouiw0YwQ0TRZivk-NjqVpT8eLZdkMMEjCROFCBRQk49LJIazz1s+TCLg+9jSgU1OnNJKly0lC3L0tsyMM1C3Q8zChxSuIbLw68srvBdEMq0qKPc9DDy8-t6JoNw3CAA)

it was fixed in 4.3.x
